tests/lib/rules/no-redeclare.js

```diff
@@ -739,3 +739,592 @@ ruleTester.run("no-redeclare", rule, {
+	  `,
+		// Valid TypeScript declaration combinations
+		`
+  interface A {}
+  interface A {}
+	  `,
+		`
+  interface A {}
+  class A {}
+	  `,
+		`
+  class A {}
+  namespace A {}
+	  `,
+		`
+  interface A {}
+  class A {}
+  namespace A {}
+	  `,
+		`
+  function A() {}
+  namespace A {}
+	  `,
+		`
+  function A() {}
+  class A {}
+	  `,
+		`
+  function A() {}
+  class A {}
+  namespace A {}
+	  `,
+		`
+  type something = string;
+  const something = 2;
+	  `,
+		`
+  declare function a(): void;
+  declare function a(): void;
+
+  declare function b(): void;
+  declare class b { }
+
+  enum c { }
+  namespace c { }
+
+  declare enum d { }
+  interface e { }
+	`,
+	],
+	invalid: [
+		{
+			code: `
+	  var a = 3;
+	  var a = 10;
+			`,
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  switch (foo) {
+		case a:
+		  var b = 3;
+		case b:
+		  var b = 4;
+	  }
+			`,
+			errors: [
+				{
+					data: {
+						id: "b",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = 3;
+	  var a = 10;
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = {};
+	  var a = [];
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  function a() {}
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  function a() {}
+	  function a() {}
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = function () {};
+	  var a = function () {};
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = function () {};
+	  var a = new Date();
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = 3;
+	  var a = 10;
+	  var a = 15;
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var a;
+			`,
+			languageOptions: { parserOptions: { sourceType: "module" } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  export var a;
+	  var a;
+			`,
+			languageOptions: { parserOptions: { sourceType: "module" } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: "var Object = 0;",
+			options: [{ builtinGlobals: true }],
```

Comment:
This case is valid in JavaScript. I think it should remain valid in TypeScript ([Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBbXCJlcnJvclwiLCB7IFwiYnVpbHRpbkdsb2JhbHNcIjogdHJ1ZSB9XSovXG5cbnZhciBPYmplY3QgPSAwO1xuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)).